### DISCUSSION TITLE
Add way to automatically inject <link ref="modulepreload">

### DISCUSSION THREAD

**Discussion by jonyo** [withastro/roadmap#561] [Proposal]
### Body

# Summary

Extend the existing `client:` options, to add a `client:preload` option for any astro island tags. Doing so would make it automatically inject `<link ref="modulepreload" src="path/to/component.hash.js">` right before the `</head>` in the markup.

That, or could adjust the `client:load` behavior to do this automatically.

# Background & Motivation

We are wanting to add `rel=modulepreload` links to the head section. This lets the browser start downloading JS in the background while the rest of the page is still loading. The problem is that the component JS files have a hash added to the filename, and we want to keep that turned on for the cache-busting reasons.

There is a work-around, you can scan the dist folder at run time, find the file that matches `expected.*.js`, and use that for the filename. But this is a lot of extra work, it would be ideal if this ability was baked in to astro.

# Goals

- Currently, no easy way to add ref=modulepreload for static JS files.
- Those JS files do not even start download until after everything else, slowing down time to interactivity.

# Example

```jsx
<html>
<head>

{/* this next line would be injected by astro */}
<link ref="modulepreload" src="_astro/targetcomponent.abc123.js">
</head>
<body>
<targetcomponent client:load client:preload foo={bar}></targetcomponent>
</body>
```

**Comment by lilnasy** [comment]
- I have started to explore this idea here: https://github.com/withastro/astro/pull/8804

**Comment by lilnasy** [comment]
- WIP PR: https://github.com/withastro/astro/pull/8804

# Summary

Astro has complete information about the module graph of client-side scripts. This can be leveraged to add `<link rel="preload">` tags allowing users to optimize their websites for first interaction.

# Example
```ts
export default defineConfig({
    preload: {
        modules: true
    }
})
```

# Goals
- Project-level opt-in configuration to preload scripts

# Non-goals
- Preloading by default
- Per-component opt-in
- Per-page opt-in

# Unresolved questions
- Should CSS be preloaded? I was not able to make astro emit a css file that imports another css file. Maybe a processing step simplifies it.
- Could fonts be preloaded? For example, the browser picks the necessary font from this file based on usage: [fontsource/ibm-plex-mono](https://github.com/fontsource/font-files/blob/d936f37/fonts/google/ibm-plex-mono/400.css).

# Related
- https://github.com/withastro/roadmap/discussions/609

**Comment by Bug-Reaper** [comment]
This seems like an important feature. The PR seems to not have been merged and the bot comment on it doesn't make sense to me.

Any update on this?

**Reply by ADTC** [reply]
@lilnasy why was the PR closed without merging?

**Reply by lilnasy** [reply]
Lack of signal from users and interest from maintainers, who were also swamped at the time

**Comment by ADTC** [comment]
I spent some 25 to 28 bucks USD on a back-and-forth with Claude Opus 4.6 to come up with [a hack that does this via Astro integration.](https://gist.github.com/ADTC/56bdcd863ff388cd9409dcc3a89ed41b) Agent-tested thoroughly and self-tested in clean non-cached server deployments to work reliably.

If it helps you, leave an upvote. Thanks. 🙂 

Note that this walks the tree for dynamic imports as well, which is probably what you want anyway. (No harm preloading them.)
